### ISSUE TITLE
SequentialLR scheduler, bug with last_epoch = N

### ISSUE THREAD

**Issue by Jostarndt** [pytorch/pytorch#162591]
### 🐛 Describe the bug

Hi! 

I observed what I believe is a bug after restarting a training. 

I am using SequentialLR and wanted to restart training after some epochs, but the learning rate was quite weird to me. After training for N steps, I started the training again and loaded a saved model, but the learning rate for which I gave last_epoch=N was completely not the one where it has stopped.

I did try to provide a a minimum example that reproduces the bug, and its behaviour is weird. Please try it both with `scheduler2 = CosineAnnealingLR(optimizer, T_max = 10)` and `scheduler2 = CosineAnnealingLR(optimizer, T_max = 10)`.

```
import torch
from torch.optim import Adam
from torch.optim.lr_scheduler import ConstantLR, ExponentialLR, SequentialLR, CosineAnnealingLR


# Create a simple model and optimizer for testing
model = torch.nn.Linear(1, 1)
optimizer = Adam(model.parameters(), lr=1.0)

# Set up schedulers
scheduler1 = ConstantLR(optimizer, factor=0.1, total_iters=20)
scheduler2 = ExponentialLR(optimizer, gamma=0.9)
scheduler2 = CosineAnnealingLR(optimizer, T_max = 10-3)

scheduler = SequentialLR(
    optimizer,
    schedulers=[scheduler1, scheduler2],
    milestones=[3],
)


# Test the scheduler
for epoch in range(10):
    current_lr = optimizer.param_groups[0]['lr']
    print(f"Epoch {epoch:2d}: LR = {current_lr:.6f}")
    scheduler.step()

print("---------Restarting from step 8: ")

scheduler = SequentialLR(
    optimizer,
    schedulers=[scheduler1, scheduler2],
    milestones=[3],
    last_epoch=8
)

for epoch in range(3):
    current_lr = optimizer.param_groups[0]['lr']
    print(f"Epoch ?: LR = {current_lr:.6f}")
    scheduler.step()
```

E.G. for `scheduler2 = CosineAnnealingLR(optimizer, T_max = 10)` in the second setting (restarting the training) the output is `1.000000, 0.000000, 0.049516`; but it should not at all start at 1.0, but either at `0.388740`or `0.188255` (as shown in the outputs before) and should definitely not go beyond 0. 

The overall output (also there is some version-issue internally I believe, thus the second warning):

```
Epoch  0: LR = 0.100000
/XXXXXX/test_env/lib/python3.10/site-packages/torch/optim/lr_scheduler.py:192: UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
  warnings.warn(
Epoch  1: LR = 0.100000
Epoch  2: LR = 0.100000
/XXXXXX/test_env/lib/python3.10/site-packages/torch/optim/lr_scheduler.py:209: UserWarning: The epoch parameter in `scheduler.step()` was not necessary and is being deprecated where possible. Please use `scheduler.step()` to step the scheduler. During the deprecation, if epoch is different from None, the closed form is used instead of the new chainable form, where available. Please open an issue if you are unable to replicate your use case: https://github.com/pytorch/pytorch/issues/new/choose.
  warnings.warn(EPOCH_DEPRECATION_WARNING, UserWarning)
Epoch  3: LR = 1.000000
Epoch  4: LR = 0.950484
Epoch  5: LR = 0.811745
Epoch  6: LR = 0.611260
Epoch  7: LR = 0.388740
Epoch  8: LR = 0.188255
Epoch  9: LR = 0.049516
---------Restarting from step 8:
Epoch ?: LR = 1.000000
Epoch ?: LR = 0.000000
Epoch ?: LR = 0.049516
```

### Versions

PyTorch version: 2.9.0.dev20250813+cu129
Is debug build: False
CUDA used to build PyTorch: 12.9
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.2) 11.4.0
Clang version: Could not collect
CMake version: version 3.27.5
Libc version: glibc-2.35

Python version: 3.10.16 (main, Dec 11 2024, 16:24:50) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-153-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.7.64
CUDA_MODULE_LOADING set to:
GPU models and configuration:
GPU 0: NVIDIA A100-SXM4-40GB
GPU 1: NVIDIA A100-SXM4-40GB
GPU 2: NVIDIA A100-SXM4-40GB
GPU 3: NVIDIA A100-SXM4-40GB

Nvidia driver version: 575.51.03
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A                                                                                                                                                                                 [75/536]
Is XNNPACK available: True

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           48 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  128
On-line CPU(s) list:                     0-127
Vendor ID:                               AuthenticAMD
Model name:                              AMD EPYC 7543 32-Core Processor
CPU family:                              25
Model:                                   1
Thread(s) per core:                      2
Core(s) per socket:                      32
Socket(s):                               2
Stepping:                                1
Frequency boost:                         enabled
CPU max MHz:                             2800.0000
CPU min MHz:                             1500.0000
BogoMIPS:                                5589.63
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nop
l nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignss
e 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 invpcid cq
m rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save ts
c_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca
Virtualization:                          AMD-V
L1d cache:                               2 MiB (64 instances)
L1i cache:                               2 MiB (64 instances)
L2 cache:                                32 MiB (64 instances)
L3 cache:                                512 MiB (16 instances)
NUMA node(s):                            8
NUMA node0 CPU(s):                       0-7,64-71
NUMA node1 CPU(s):                       8-15,72-79
NUMA node2 CPU(s):                       16-23,80-87
NUMA node3 CPU(s):                       24-31,88-95
NUMA node4 CPU(s):                       32-39,96-103                                                                                                                                                       [38/536]
NUMA node5 CPU(s):                       40-47,104-111
NUMA node6 CPU(s):                       48-55,112-119
NUMA node7 CPU(s):                       56-63,120-127
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Mitigation; safe RET
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Retpolines; IBPB disabled; IBRS_FW; STIBP disabled; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                     Not affected
Vulnerability Tsx async abort:           Not affected

Versions of relevant libraries:
[pip3] numpy==2.1.2
[pip3] nvidia-cublas-cu12==12.9.1.4
[pip3] nvidia-cuda-cupti-cu12==12.9.79
[pip3] nvidia-cuda-nvrtc-cu12==12.9.86
[pip3] nvidia-cuda-runtime-cu12==12.9.79
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.4.1.4
[pip3] nvidia-curand-cu12==10.3.10.19
[pip3] nvidia-cusolver-cu12==11.7.5.82
[pip3] nvidia-cusparse-cu12==12.5.10.65
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nvjitlink-cu12==12.9.86
[pip3] nvidia-nvtx-cu12==12.9.79
[pip3] pytorch-triton==3.4.0+gitf7888497
[pip3] torch==2.9.0.dev20250813+cu129
[pip3] torchmetrics==1.8.1
[pip3] torchvision==0.24.0.dev20250813+cu129
[pip3] triton==3.4.0
[conda] blas                      1.0                         mkl
[conda] cuda-cudart               11.8.89                       0    nvidia
[conda] cuda-cupti                11.8.87                       0    nvidia
[conda] cuda-libraries            11.8.0                        0    nvidia
[conda] cuda-nvrtc                11.8.89                       0    nvidia
[conda] cuda-nvtx                 11.8.86                       0    nvidia
[conda] cuda-runtime              11.8.0                        0    nvidia
[conda] cudatoolkit               11.5.1               hcf5317a_9    nvidia
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] intel-openmp              2023.1.0         hdb19cb5_46305
[conda] libcublas                 11.11.3.6                     0    nvidia
[conda] libcufft                  10.9.0.58                     0    nvidia
[conda] libcurand                 10.3.3.141                    0    nvidia
[conda] libcusolver               11.4.1.48                     0    nvidia
[conda] libcusparse               11.7.5.86                     0    nvidia
[conda] mkl                       2023.1.0         h213fc3f_46343
[conda] mkl-service               2.4.0            py39h5eee18b_1
[conda] mkl_fft                   1.3.8            py39h5eee18b_0
[conda] mkl_random                1.2.4            py39hdb19cb5_0
[conda] numpy                     1.25.2           py39h5f9d8c6_0
[conda] numpy-base                1.25.2           py39hb5e798b_0
[conda] nvidia-cublas-cu11        11.10.3.66               pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu11    11.7.99                  pypi_0    pypi
[conda] nvidia-cuda-runtime-cu11  11.7.99                  pypi_0    pypi
[conda] nvidia-cudnn-cu11         8.5.0.96                 pypi_0    pypi
[conda] pyg                       2.3.1           py39_torch_2.0.0_cu118    pyg
[conda] pytorch                   2.0.1           py3.9_cuda11.8_cudnn8.7.0_0    pytorch
[conda] pytorch-cuda              11.8                 h7e8668a_5    pytorch
[conda] pytorch-lightning         2.0.9              pyhd8ed1ab_0    conda-forge
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] tbb                       2021.8.0             hdb19cb5_0
[conda] torch                     1.13.1                   pypi_0    pypi
[conda] torchaudio                2.0.2                py39_cu118    pytorch
[conda] torchmetrics              0.11.0                   pypi_0    pypi
[conda] torchtriton               2.0.0                      py39    pytorch
[conda] torchvision               0.15.2               py39_cu118    pytorch
